3.0.0: Convert algod responses to typed

Eric-Warehime opened the pull request:
Resolves #771 for the algod client. 

I'm attempting to merge this to the `3.0.0` branch of the SDK which we _will not_ be releasing until we've gained a critical mass of breaking changes that we feel justifies a breaking major version bump for this SDK.

Also updates the eslint dependencies to more recent versions, and fixes a couple of new lints which were failing--optional parameters at the end of the function signature, and promise return in the cucumber tests.

Existing tests pass--not sure if we need to be adding more tests here.

jasonpaulos commented on src/client/v2/algod/accountApplicationInformation.ts:
```diff
@@ -1,8 +1,9 @@
+import { AccountApplicationResponse } from './models/types';
 
-export default class AccountApplicationInformation extends JSONRequest {
+export default class AccountApplicationInformation extends JSONRequest<AccountApplicationResponse> {
```
These changes only affect the typescript types that the `do` method says it returns -- no actual code change has happened.

To properly implement the change, you need to inherit from `JSONRequest<Model, Record<string, any>>` **and** override the `prepare(Record<string, any>): Model` method.

Eric-Warehime commented:
Requires https://github.com/algorand/algorand-sdk-testing/pull/299 to be merged for cucumber tests to start passing.

algochoi commented on src/client/v2/algod/sendRawTransaction.ts:
```diff
@@ -49,7 +50,17 @@ export default class SendRawTransaction extends JSONRequest {
-    return res.body;
+    const res = await this.c.post(
+      this.path(),
+      Buffer.from(this.txnBytesToPost),
```
I think `Buffer.from` conversions are unnecessary now:
```suggestion
      this.txnBytesToPost,
```

Eric-Warehime commented on tests/cucumber/steps/steps.js:
```diff
@@ -180,6 +180,24 @@ module.exports = function getSteps(options) {
     return boxRefArray;
   }
 
+  let doRaw = false;
```
I'm not sure if I put it on every one, but most of them. Specifically because any step that's along the lines of `and we make a XYZ call` either already has mocking setup and a `do` call will pass, or it is going to return random stuff so we need to use `doRaw` otherwise it will throw an exception.

Eric-Warehime commented on tests/cucumber/steps/steps.js:
```diff
@@ -180,6 +180,24 @@ module.exports = function getSteps(options) {
 
+  let doRaw = false;
+
+  function doOrDoRaw(req) {
```
We could do this if each step that made a call knew whether or not it should expect a valid response. To do that we either need to modify the step definition (doesn't make sense just to work for JS), or we can override the step like we do for some steps in this block https://github.com/algorand/js-algorand-sdk/blob/develop/tests/cucumber/steps/index.js#L436

However, since they're each a separate step definition we'd either need to use some string matching to tell, or override it separately for every API call. This global `doOrDoRaw` seemed the best option to me, but I'm open to suggestions on how to handle this better.

algochoi replied:
I see now. I also took a look and don't have a better suggestion, but it wasn't immediately obvious why we had it, so a [comment like we have here](https://github.com/algorand/js-algorand-sdk/blob/bc272cb8048e9cf1ae3b311e2ac17da2ae35ff0f/tests/cucumber/steps/steps.js#L95) might be nice to explain the context of the global variable.

algochoi commented on src/wait.ts:
```diff
@@ -13,15 +14,15 @@ export async function waitForConfirmation(
     throw new Error('Unable to get node status');
   }
-  const startRound = status['last-round'] + 1;
+  const startRound = Number(status.lastRound) + 1;
```
We lose precision by converting to a `Number` as rounds are uint64s and the max `Number` is ~2^53. This might cause overflow on things like dev networks, so it seems prudent to use `BigInt` where we can in source code. 

It also seems problematic that our existing APIs take `number` instead of `BigInt` for round numbers... https://github.com/algorand/js-algorand-sdk/blob/bc272cb8048e9cf1ae3b311e2ac17da2ae35ff0f/src/client/v2/algod/block.ts#L11
I think we should change these APIs to accept `BigInt` as well?

Eric-Warehime commented on src/makeTxn.ts:
```diff
@@ -370,8 +370,8 @@ export function makeAssetConfigTxnWithSuggestedParams(
   suggestedParams: MustHaveSuggestedParams<AssetConfigTxn>['suggestedParams'],
-  strictEmptyAddressChecking = true,
-  rekeyTo?: AssetConfigTxn['reKeyTo']
+  rekeyTo?: AssetConfigTxn['reKeyTo'],
```
My eslint complains that defaulted arguments should come last which is why I made this change.

algochoi reviewed (approved):
LGTM 👍 

I don't feel as strongly in converting everything to BigInt's anymore, but I think that applying a consistent `number | bigint` type for the API args is a win for v3.

jasonpaulos reviewed (commented):
Nice job, this is going to be a huge improvement for developers.

Higher level comments:
1. What about indexer responses? Is that on the roadmap for v3 as well?
2. Our generated response models use the type `number | bigint` for all integers. I find this pretty frustrating, since you can't really do anything without converting to one of these types. In comparison, for binary data we accept `string | Uint8Array`, but convert everything to `Uint8Array` before storing in a field. I think we should do a similar thing for integers, but use `bigint` since it's the most general type.
3. We should definitely change the default integer decoding behavior in this release. See [parseJSON](https://github.com/algorand/js-algorand-sdk/blob/4e1073d6a15b443a73dedf42b025173b397624b0/src/utils/utils.ts#L27) for more info, but basically we default to `IntDecoding.DEFAULT`, which means all types are decoded as `number`s, which isn't accurate for values above 2^53-1. Since the response types declare they can handle `bigint`s, let's actually start populating them with `bigint`s by default.

Eric-Warehime commented:
@jasonpaulos 
Ok I think I've addressed all of the feedback except for the following:
1. Indexer response typing--I've not included them in this PR. It's possible we should include them in this branch before releasing 3.0.0 since it would be a breaking change as well.
2. Converting the storage format of everything to bigint instead of union types. This seems both like a breaking change to me, and also something that's more of an annoyance than anything else. Libraries like https://www.npmjs.com/package/json-bigint seem to solve this problem.

Eric-Warehime commented on tests/cucumber/steps/steps.js:
```diff
@@ -2,6 +2,7 @@
 const assert = require('assert');
 const fs = require('fs');
 const path = require('path');
+const JSONBig = require('json-bigint');
```
Confused because we have this though https://github.com/algorand/js-algorand-sdk/blob/develop/package.json#L31

Eric-Warehime replied:
Also, I only see those failures in firefox tests. The node tests hit
```
fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/community/x86_64/APKINDEX.tar.gz
#7 0.441 ERROR: unable to select packages:
#7 0.442   bash (no such package)
```

and the chrome tests hit
```
SessionNotCreatedError: session not created: This version of ChromeDriver only supports Chrome version 114
Current browser version is 116.0.5845.96 with binary path /usr/bin/google-chrome
    at Object.throwDecodedError (/home/circleci/project/node_modules/selenium-webdriver/lib/error.js:524:15)
    at parseHttpResponse (/home/circleci/project/node_modules/selenium-webdriver/lib/http.js:601:13)
    at Executor.execute (/home/circleci/project/node_modules/selenium-webdriver/lib/http.js:529:28)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

algochoi replied:
Locally, node seems to work but is a little flaky upon startup. Our chromedriver package needs a bump, and that change should probably be merged into master: https://github.com/algorand/js-algorand-sdk/pull/811

jasonpaulos commented on tests/cucumber/steps/steps.js:
```diff
@@ -121,6 +121,10 @@ module.exports = function getSteps(options) {
 
+  // We need to import inside the steps export so that the browser tests have the dependency imported
+  // eslint-disable-next-line global-require
+  const JSONBig = require('json-bigint');
```
I'm not sure if this will work. However we do have the `tests/cucumber/browser/test.js` file which can be used to populate imports for browser tests.

Adding something like this to that file should work (we do the same for the `assert` package):

```javascript
const JSONBig = require('json-bigint');
window.JSONBig = JSONBig;
```

jasonpaulos commented:
@Eric-Warehime I realized changing the default int decoding is a bit of a larger problem. I've opened https://github.com/algorand/js-algorand-sdk/pull/816 to hopefully address the bigint issue.

If the idea behind that PR makes sense to you, this PR can revert the attempt to change int decoding in favor of that PR.

Everything else here looks good to me.

jasonpaulos commented on src/client/v2/algod/block.ts:
```diff
@@ -6,12 +6,10 @@ import HTTPClient from '../../client';
  * block gets the block info for the given round. this call may block
  */
 export default class Block extends JSONRequest {
```
Why doesn't this one return `BlockResponse`?

Eric-Warehime replied:
I'm not convinced it would be correct. The query uses msgp format which means the response will be the msgp-encoded raw block bytes.

jasonpaulos replied:
We use msgpack for a few other endpoints, like `PendingTransactions`, so I think that's something that can be handled.

What I see is a little different about this response is that its generated model contains a reference to `BlockHeader`, which is a handwritten definition in this SDK. I think this definition was typed based on the assumption that it would be JSON decoded, not msgpack decoded, so some of its types are not accurate -- e.g. msgpack binary data decodes as `Uint8Array` but there it's defined as `string`. Maybe this is what you were referring to?

I'd really like for us to return `BlockResponse` for this request as well, both for consistency and because it will make it much easier to make non-breaking improvements to the response type over time.

So I'd like to propose two options here:
1. Return `BlockResponse` and update our handwritten `BlockHeader` to have accurate types when decoded from msgpack.
2. Return `BlockResponse`, but change our handwritten `BlockHeader` to be a generic `Record<string, any>`. In the future we have the option of making this type more specific, which I don't think would necessarily be a breaking change. And even if we don't, users get some typing benefits (the top level response would be typed, with the two properties, `block` and `cert`).

Eric-Warehime replied:
Tests pass for me when doing https://github.com/algorand/js-algorand-sdk/pull/776/commits/fef19974328ab7a31fdc47f24cd7585fb863025f 

Let me know what you think.

Eric-Warehime commented:
@jasonpaulos 

If you want me to I can remove the bigint changes in this PR. I'm not super opinionated on having `number | bigint` unions w/ the deserialization prefering bigints versus what it looks like you've done in your PR which is to just convert all numbered responses to bigint exclusively. 

I guess the caveat there is that you'll need to alter the code generator since you're modifying the types and it looks like your PR has some issues (possibly some of the same test failures I've already worked through here?). 

As a final option I guess we could get this merged and follow up with a PR to change the unions to bigint--maybe that would just cause a lot of unnecessary work though.

jasonpaulos reviewed (approved):
I still think our `BlockHeader` needs to be updated to be correct, but that can be worked out later. Let's get this merged
